src/utils/jwt.ts

```diff
@@ -0,0 +1,240 @@
+    if (!validator.isURL(clientCertUrl)) {
+      throw new Error(
+        'The provided public client certificate URL is an invalid URL.',
+      );
+    }
+  }
+
+  /**
+   * Fetches the public keys for the Google certs.
+   *
+   * @return A promise fulfilled with public keys for the Google certs.
+   */
+  public fetchPublicKeys(): Promise<{ [key: string]: string }> {
+    if (this.shouldRefresh()) {
+      return this.refresh();
+    }
+    return Promise.resolve(this.publicKeys);
+  }
+
+  private shouldRefresh(): boolean {
+    const publicKeysExist = (typeof this.publicKeys !== 'undefined');
+    const publicKeysExpiredExists = (typeof this.publicKeysExpireAt !== 'undefined');
+    const publicKeysStillValid = (publicKeysExpiredExists && Date.now() < this.publicKeysExpireAt);
+    return !(publicKeysExist && publicKeysStillValid);
+  }
+
+  private refresh(): Promise<{ [key: string]: string }> {
+    const client = new HttpClient();
+    const request: HttpRequestConfig = {
+      method: 'GET',
+      url: this.clientCertUrl,
+      httpAgent: this.httpAgent,
+    };
+    return client.send(request).then((resp) => {
+      if (!resp.isJson() || resp.data.error) {
+        // Treat all non-json messages and messages with an 'error' field as
+        // error responses.
+        throw new HttpError(resp);
+      }
+      if (Object.prototype.hasOwnProperty.call(resp.headers, 'cache-control')) {
+        const cacheControlHeader: string = resp.headers['cache-control'];
+        const parts = cacheControlHeader.split(',');
+        parts.forEach((part) => {
+          const subParts = part.trim().split('=');
+          if (subParts[0] === 'max-age') {
+            const maxAge: number = +subParts[1];
+            this.publicKeysExpireAt = Date.now() + (maxAge * 1000);
+          }
+        });
+      }
+      this.publicKeys = resp.data;
+      return resp.data;
+    }).catch((err) => {
+      if (err instanceof HttpError) {
+        let errorMessage = 'Error fetching public keys for Google certs: ';
+        const resp = err.response;
+        if (resp.isJson() && resp.data.error) {
+          errorMessage += `${resp.data.error}`;
+          if (resp.data.error_description) {
+            errorMessage += ' (' + resp.data.error_description + ')';
+          }
+        } else {
+          errorMessage += `${resp.text}`;
+        }
+        throw new JwtError(JwtErrorCode.INTERNAL_ERROR, errorMessage);
+      }
+      throw err;
+    });
+  }
+}
+
+/**
+ * Verifies JWT signature with a public key.
+ */
+export class PublicKeySignatureVerifier implements SignatureVerifier {
+  constructor(private keyFetcher: UrlKeyFetcher) {
+    if (!validator.isNonNullObject(keyFetcher)) {
+      throw new Error('The provided key fetcher is not an object or null.');
+    }
+  }
+
+  public verify(token: string): Promise<void> {
+    const ALGORITHM_RS256: jwt.Algorithm = 'RS256' as const;
+    const error = new JwtError(JwtErrorCode.INVALID_TOKEN,
```

Comment:
This bit is duplicated in the other implementation. If you change the `isSignatureValue()` helper to not return a boolean, and instead signal errors by rejecting, you can avoid this. Given a signature like:

```
function verifyJwtSignature(token, getKey, options): Promise<void>
```

this will become:

```
return verifyJwtSignature(token, getKeyCallback(this.keyFetcher), {algorithms: [RS_256]});
```